libnetwork/osl/interface_linux.go

```diff
@@ -276,19 +288,104 @@ func (n *Namespace) AddInterface(ctx context.Context, srcName, dstPrefix string,
 	if err != nil {
 		return fmt.Errorf("failed to set link up: %v", err)
 	}
+	log.G(ctx).Debug("link has been set to up")
 
 	// Set the routes on the interface. This can only be done when the interface is up.
 	if err := setInterfaceRoutes(ctx, nlh, iface, i); err != nil {
 		return fmt.Errorf("error setting interface %q routes to %q: %v", iface.Attrs().Name, i.Routes(), err)
 	}
 
+	// When the net.ipv4.conf.[iface].arp_notify sysctl is set to 1, the
+	// kernel is supposed to send a gratuitous ARP when the interface is
+	// brought up. However, this functionality is broken on at least the
+	// Ubuntu 20.04 kernel
+	// https://bugs.launchpad.net/ubuntu/+source/linux/+bug/1989187
+	// and Docker Desktop's 6.6.16-linuxkit kernel. One option would be to
+	// send the gratuitous ARP ourselves from userspace using a raw socket,
+	// but that would not respect the arp_notify sysctl setting. Instead,
+	// we will trigger the kernel to send a gratuitous ARP using the other
+	// event: setting the interface's MAC address after the kernel has
+	// finished bringing up the link asynchronously.
+	// Do this even if there's no IPv4 address, it also means an IPv6
+	// Neighbour Advertisement is sent immediately.
+	<-upped
+	err = func() error {
+		iface, err := nlh.LinkByIndex(iface.Attrs().Index)
+		if err != nil {
+			return fmt.Errorf("failed to refresh link attributes: %w", err)
+		}
+		log.G(ctx).WithFields(log.Fields{
+			"iface": iface.Attrs().Name,
+			"ifi":   iface.Attrs().Index,
+			"mac":   iface.Attrs().HardwareAddr,
+			"ip":    i.Address(),
+			"ip6":   i.AddressIPv6(),
+		}).Debug("applying gratuitous ARP workaround")
+		return setInterfaceMAC(ctx, nlh, iface, i)
+	}()
+	if err != nil {
+		log.G(ctx).WithError(err).Warn("could not apply gratuitous ARP workaround")
+	}
+
 	n.mu.Lock()
 	n.iFaces = append(n.iFaces, i)
 	n.mu.Unlock()
 
 	return nil
 }
 
+func waitForIfUpped(ctx context.Context, ns netns.NsHandle, ifIndex int) (chan struct{}, error) {
+	update := make(chan netlink.LinkUpdate, 100)
+	upped := make(chan struct{})
+	opts := netlink.LinkSubscribeOptions{
+		ListExisting: true, // in case the link has NO_CARRIER, so no events are sent
+		ErrorCallback: func(err error) {
+			log.G(ctx).WithFields(log.Fields{
+				"ifi":   ifIndex,
+				"error": err,
+			}).Debug("netlink error while waiting for interface up")
```

Comment:
As discussed on Slack, I saw this error while testing this PR (second-to-last log):

```
DEBU[2024-11-06T10:11:28.907169337Z] link has been set to up                       spanID=ec4fd13878d1114f traceID=e489e9906a34cf8491e39f49a60d219b
DEBU[2024-11-06T10:11:28.907424671Z] link update                                   flags="deviceFlags(IFF_UP | IFF_BROADCAST | IFF_MULTICAST | IFF_LOWER_UP)" iface=eth0 ifi=14 spanID=ec4fd13878d1114f traceID=e489e9906a34cf8491e39f49a60d219b
DEBU[2024-11-06T10:11:28.907488212Z] link update                                   flags="deviceFlags(IFF_UP | IFF_BROADCAST | IFF_MULTICAST | IFF_LOWER_UP)" iface=eth0 ifi=14 spanID=ec4fd13878d1114f traceID=e489e9906a34cf8491e39f49a60d219b
DEBU[2024-11-06T10:11:28.907534337Z] link update                                   flags="deviceFlags(IFF_UP | IFF_BROADCAST | IFF_RUNNING | IFF_MULTICAST | IFF_LOWER_UP)" iface=eth0 ifi=14 spanID=ec4fd13878d1114f traceID=e489e9906a34cf8491e39f49a60d219b
DEBU[2024-11-06T10:11:28.907587671Z] netlink error while waiting for interface up  error="Receive failed: resource temporarily unavailable" ifi=14 spanID=ec4fd13878d1114f traceID=e489e9906a34cf8491e39f49a60d219b
DEBU[2024-11-06T10:11:28.907681129Z] applying gratuitous ARP workaround            iface=eth0 ifi=14 ip=172.19.0.2/16 ip6="fd9b:a045:3ddd::2/64" mac="02:42:ac:13:00:02" spanID=ec4fd13878d1114f traceID=e489e9906a34cf8491e39f49a60d219b
```

After investigating, we found out that this is totally benign. What happens is that the `defer close(upped)` below is used to signal to `LinkSubscribeWithOptions` it should stop watch the interface. But then, in vishvananda/netlink these two goroutines are racing against each other:

- https://github.com/vishvananda/netlink/blob/main/link_linux.go#L2435-L2438
- https://github.com/vishvananda/netlink/blob/main/link_linux.go#L2453-L2459

That second goroutine doesn't special-case errors induced by the `s.Close()` call made by the 1st goroutine.

So either we need to fix vishvananda (IMO sending this EAGAIN is really fishy!), or we need to add another defer running before we close the channel to signal to `ErrorCallback` that it shouldn't log anything anymore.

Reply:
That's done - thank you ... I also added a check for the `update` channel closing early, in case the netlink socket read fails for some other reason, causing the function in the netlink package to bail out before the `done` channel is closed.